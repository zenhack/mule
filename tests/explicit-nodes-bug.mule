# This tests for a regression. The original bug was that the below would
# erroneously fail to type check, because in the return value of
# from-order, the constraint graph ends up with an equality constraint
# on something bound inside the body of the type lambda, which fails
# due to explicit binding permissions.
#
# We want references like ord.t inside the body of lambdas to be bound
# outside of it.
let
  , type sig k = {
      , type t v
    }
  , from-order : all a. { type t = a } -> sig a = fn ord.
    {
      , type t v = ord.t
    }
in
0
