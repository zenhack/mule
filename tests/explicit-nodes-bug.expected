
Desugared:

	(letrec
	 (types
	  (((sig (lam k (rec sig (exist $6 (record (types ((t $6))) (values ())))))))))
	 (values
	  ((from-order
	    (:
	     (fn ord (record (types (((t (lam v (rec t (. ord t))))))) (values ())))
	     (all a ((record (types ((t (rec t a)))) (values ())) -> (sig a)))))))
	 (Integer 0))

inferred type:

	int

Runtime term:

	(LetRec (((0 (Lam (0 () (Record ()))))) (Const (Integer 0))))

Evaluated:

	(Const (Integer 0))
